006-securing-kafka-connect-rest-endpoint.md

```diff
@@ -0,0 +1,146 @@
+ - https://github.com/strimzi/strimzi-kafka-operator/issues/3229
+ - https://github.com/strimzi/strimzi-kafka-operator/issues/2161
+ - https://cloud-native.slack.com/archives/CMH3Q3SNP/p1577722052135000
+
+## Proposed changes
+
+This proposal adds a `rest` property to the KafkaConnect spec, allowing users to configure Connect REST API  endpoints that use HTTP or HTTPS. Client authentication is not addressed here, but there is scope for future proposals to address it.
+
+Simple example CR:
+
+```
+apiVersion: kafka.strimzi.io/v1beta1
+kind: KafkaConnect
+spec:
+  rest:
+  - tls: {}
+  ...
+```
+
+This CR would generate a cert and key in a secret (for use by the operator clients), create SSL truststores and keystores from the cert and key and set the following properties in the generated connect configuration to enable an HTTPS connection:
+
+```
+listeners: https://:8443
+rest.advertised.listener: https
+rest.advertised.port: 8443
+
+listeners.https.ssl.client.auth: none
+listeners.https.ssl.truststore.location: /tmp/kafka/kafka-connect-rest.truststore.p12
+listeners.https.ssl.truststore.password: ***generated password***
+listeners.https.ssl.truststore.type: PKCS12
+listeners.https.ssl.keystore.location: /tmp/kafka/kafka-connect-rest.keystore.p12
+listeners.https.ssl.keystore.password: ***generated password***
+listeners.https.ssl.keystore.type: PKCS12
+
+```
+
+A more complex example CR:
+
+```
+apiVersion: kafka.strimzi.io/v1beta1
+kind: KafkaConnect
+spec:
+  rest:
+  - port: 8083
+  - port: 8483
+    tls:
+      configuration:
+        brokerCertChainAndKey:
+          secretName: mysecret
+          certificate: keystore.crt
+          key: keystore.key
+  ...
+```
+
+Here the operator configures two listeners: an unsecured HTTP listener on port 8083 and a secured HTTPS listener on port 8483 using the certificate and key from the provided secret. This uses a similar interface as used for the Kafka CR `spec.listeners.tls` property. The operator mounts the secret certificates in the Kafka Connect pods and sets a failed status and logging if the listener options clash or are mis-configured.
+
+The operator REST API client code will prefer HTTPS over HTTP when both are configured, in a similar way to how the Kafka Connect workers coordinate via HTTPS when it is available (see [KIP 507](https://cwiki.apache.org/confluence/display/KAFKA/KIP-507%3A+Securing+Internal+Connect+REST+Endpoints)).
+
+The above proposal will require modifications to the `AbstractConnectOperator` to correctly read and use the additional CR configuration, apply the TLS certs as keystore credentials to a WebClient truststore and then the webclient (or its configuration) will have to be passed into the `KafkaConnectApi` interface and `Impl` so that each call can optionally use this. This should mean for regular http calls it will continue to work as is, but the calls will also be able to speak to the secured endpoint if configured.
+
```

Comment:
Can a user configure a different network policy for the different ports? Presumably using the same mechanism as Kafka CR `spec.listeners.tls`, so `tls` has a `networkPolicyPeers`, right? Again, I think it would be good to describe the schema in a bit more detail, especially as you said 

>  This uses a _similar_ interface as used for the Kafka CR `spec.listeners.tls` property.

(my emphasis) a couple of paragraphs above.

Reply:
This proposal wasn't planning to address additional network policies beyond the default network policy that denies access to the REST port from anything other than the operator. As is currently the case, users can add network policies in k8s to allow wider access. It's a good idea though - I have added a note to the `Future extensions` section. Do you feel it's a necessary function we should be adding?

Reply:
No, I don't think it is necessary, but thanks for clarifying.